pkg/api/pluginproxy/access_token_provider_test.go

```diff
@@ -91,4 +99,217 @@ func TestAccessToken(t *testing.T) {
 			So(token2, ShouldEqual, "abc")
 		})
 	})
+
+	Convey("Plugin with token auth route", t, func() {
+		apiHandler := http.NewServeMux()
+		server := httptest.NewServer(apiHandler)
+
+		defer server.Close()
+		pluginRoute := &plugins.AppPluginRoute{
+			Path:   "pathwithtokenauth1",
+			Url:    "",
+			Method: "GET",
+			TokenAuth: &plugins.JwtTokenAuth{
+				Url: server.URL + "/oauth/token",
+				Scopes: []string{
+					"https://www.testapi.com/auth/monitoring.read",
+					"https://www.testapi.com/auth/cloudplatformprojects.readonly",
+				},
+				Params: map[string]string{
+					"grant_type":    "client_credentials",
+					"client_id":     "{{.JsonData.client_id}}",
+					"client_secret": "{{.SecureJsonData.client_secret}}",
+					"audience":      "{{.JsonData.audience}}",
+					"client_name":   "datasource_plugin",
+				},
+			},
+		}
+
+		templateData := templateData{
+			JsonData: map[string]interface{}{
+				"client_id": "my_client_id",
+				"audience":  "www.example.com",
+			},
+			SecureJsonData: map[string]string{
+				"client_secret": "my_secret",
+			},
+		}
+
+		ds := &models.DataSource{Id: 1, Version: 2}
+
+		var i int
+		apiHandler.HandleFunc("/oauth/token", func(w http.ResponseWriter, req *http.Request) {
+			json.NewEncoder(w).Encode(token)
+			i++
+		})
+
+		Convey("Should cache token", func() {
+			Convey("Should parse and cache access token with expires_in string field", func() {
```

Comment:
It's a bit difficult to understand what are the different test cases and there is too much repeated code. I would suggest to define an array of structs which describe each test like:
```
struct {
				desc              string
				expiresIn         interface{}
				expiresOn         interface{}
				expectedExpiresOn int64
}
```
and loop over this.